### PR TITLE
Support anchor names for inherited methods

### PR DESCRIPTION
**Pull request by sgugger**
There is a problem in the way anchors are generated for methods of classes that are inherited from a superclass. The problem was first reported in [this PR](https://github.com/huggingface/datasets/pull/3999) but it happens more broadly when you have an object (let's say `BertTokenizer`) and a method (let's say the `__call__` magic method) that you are asking in the documentation, but that method's documentation is actually written in the superclass (in our example `PreTrainedTokenizerBase`).

Currently, doc-builder will create the documentation of this method under the anchor `PreTrainedTokenizerBase.__call__` which has two problems:
- if we were in a page with two different tokenizers documenting that method, we would get the same anchor twice
- more annoying is that this method can never be linked to via inter-links. When asking for a link to this method via ``[`BertTokenizer.__call__`]``, we get a link to the `BertTokenizer.__call__` anchor

And even more insidious, the object found by `doc-builder` when looking at `BertTokenizer.__call__` is `PreTrainedTokenizerBase.__call__` (that's why this is the generated anchor name), so when resolving ``[`BertTokenizer.__call__`]``, it tries to find in the anchors mapping `PreTrainedTokenizerBase.__call__`, but in that mapping, the method is under the child's name `BertTokenizer.__call__`, so the reference is not resovled.

This PR solves those issues the following way:
- it adds an `anchor_name` to the `document_object` method, to force the anchor name to be `{documented_class_anchor}.{method}` in those situations
- if the `anchor_name` is different from the object name (as above) it saves both names in the anchor mapping
- when it's time to build links from those mappings, it creates, for the object name a link to the proper anchor name.

### CODE REVIEW

File: src/doc_builder/autodoc.py
Comment (sgugger) [inline src/doc_builder/autodoc.py:433]: This replicates what the doc-builder will do when we try to find we try to resolve link to ``[`BertTokenizer.__call__`]`` -> it will get a different `method_name` compared to the `anchor_name`.

File: src/doc_builder/autodoc.py
Comment (sgugger) [inline src/doc_builder/autodoc.py:482]: Some of the pages already have the anchor in the mapping (the ones where the object name is different from the anchor name).

File: src/doc_builder/build_doc.py
Comment (sgugger) [inline src/doc_builder/build_doc.py:217]: When we have several anchors for the same object, the first one is the real anchor name, the other ones, aliases we register.

Comment (LysandreJik) [reply src/doc_builder/build_doc.py:216]: Sounds good!

**Review by LysandreJik — APPROVED** [review]
Looks good to me!

**Comment by sgugger** [comment]
Merging this without waiting for Mishig's approval since he is on vacation, but will address any comments he has in a follow-up PR if needed :-)

**Pull request merged**
